### ISSUE TITLE
MongoDB support

### ISSUE THREAD

**Issue by AndreiMaz** [nopSolutions/nopCommerce#3737]
Similar to https://github.com/nopSolutions/nopCommerce/issues/415

**Comment by atulrungta** [comment]
Hey @AndreiMaz ,

Have your team made any progress so far on this feature? Unfortunately, two of my large sized customers went on hold when I told them that nopCommerce 4.30 will have MongoDB support for sure LOL..

I am asking this because I didn't saw any commits regarding it in last 10 days :)

**Comment by AndreiMaz** [comment]
@atulrungta Currently we're working on https://github.com/nopSolutions/nopCommerce/issues/239. Then we'll start working on other tasks

**Comment by atulrungta** [comment]
Hello @AndreiMaz ,

Are you still planning to provide this feature in current road-map?

**Comment by AndreiMaz** [comment]
@atulrungta yes

**Comment by atulrungta** [comment]
So, will you work on this feature once moving away from EF is done? I really wish to see 4.30 as a stable version with MongoDB. Almost 4 big sized projects are on hold :)

**Comment by jabberdeepa** [comment]
> So, will you work on this feature once moving away from EF is done? I really wish to see 4.30 as a stable version with MongoDB. 
@atulrungta have you try [GrandNode](https://github.com/grandnode/grandnode) project?

**Comment by atulrungta** [comment]
> > So, will you work on this feature once moving away from EF is done? I really wish to see 4.30 as a stable version with MongoDB.
> 
> @atulrungta have you try [GrandNode](https://github.com/grandnode/grandnode) project?

Yes. We worked on GrandNode for few projects in past. It is very unstable.
I strongly never recommend GrandNode unless insisted by customers :)

**Comment by atulrungta** [comment]
@AndreiMaz have you really removed this feature from 4.30 milestone?
It means, we won't see MongoDB support in this version?

**Comment by AndreiMaz** [comment]
@atulrungta After moving away from EF in this version, we can implement MongoDB quite fast. But we don't want to postpone this release. That's why it'll be implemented in 4.40. Good news is that we decided to set the release cycles again back to 6 months

**Comment by atulrungta** [comment]
> @atulrungta After moving away from EF in this version, we can implement MongoDB quite fast. But we don't want to postpone this release. That's why it'll be implemented in 4.40. Good news is that we decided to set the release cycles again back to 6 months

Hmm.. it means we have to wait for another 7 months to see MongoDB. It was most awaited thing for me in this version but yeah, good things are happening :)

**Comment by KrzysztofPajak** [comment]
@atulrungta 
Could you explain what is unstable?

**Comment by KrzysztofPajak** [comment]
@AndreiMaz 
As we talked in june, I don't think it is possible to move to MongoDB.
There is a lot of difference between nosql vs sql database.
But we will see :)
I think the NopCommerce team should focus on integration with Elasticsearch

**Comment by atulrungta** [comment]
> @atulrungta
> Could you explain what is unstable?

Are you talking about GrandNode? Their code have a lot of bugs as per last time I checked around a year back.
Their implementation of MongoDB is also slower than expected as per our load tests at that time.
The website was going back and forth on almost all of deployments we did. We fixed that with certain code changes regarding MongoDB at that time.
Although it is on .net core completely but it was still referring to .net framework at that time.
I don't know whether they have improved now, but my past experience is certainly not good enough to suggest any of my customers to use it.

**Comment by AndreiMaz** [comment]
@atulrungta Right, we plan to implement it in 4.40

**Comment by poyker** [comment]
Hi guys,

What is the point of adding yet another DB support?
Will nopCommerce users benefit from using MongoDB and what these benefits will be?

**Comment by xcellenceit** [comment]
> 
> 
> Hi guys,
> 
> What is the point of adding yet another DB support?
> Will nopCommerce users benefit from using MongoDB and what these benefits will be?

I think you're raising valid point here.

**Comment by AndreiMaz** [comment]
@poyker Increase our customer base and popularity by giving more choices to our users. Also we think that  it can significantly increase the performance compared to other databases (we'll check it very-very soon)

**Comment by mi-sohel** [comment]
@AndreiMaz thanks. Yes, it will help significantly increase performance. Is there any update on it?

**Comment by PanagLtd** [comment]
@AndreiMaz did you already plan how this is going to be implemented? Because linq2db has no support for MongoDB.

**Comment by mehmetunal** [comment]
When will mongodb support come? When Nopcommerce mongodb support comes, will you be paid or publish as opensource?

**Comment by loid345** [comment]
https://www.enterprisedb.com/news/new-benchmarks-show-postgres-dominating-mongodb-varied-workloads

**Comment by exileDev** [comment]
We are still unable to implement this task due to compatibility issues with the MongoDB driver's LINQ provider.

Currently, the driver doesn't support mappings based on interfaces (such as our ISoftDeleted filters). There are also many issues with translating expressions that use StringComparison and similar constructs. In addition, many of our complex queries that use extension methods, grouping, and other advanced LINQ features cannot be translated by the driver.

At the moment, it is unclear whether all affected queries can be adapted and how much effort such changes would require. As a result, we are currently unable to implement this feature.
